pandas/tests/util/test_util.py

```diff
@@ -417,6 +417,30 @@ def test_numpy_errstate_is_default():
     assert np.geterr() == expected
 
 
+@td.skip_if_windows
+def test_can_set_locale_invalid_set():
+    # Setting an invalid locale should return False
+    assert tm.can_set_locale('non-existent_locale') is False
+
+
+@td.skip_if_windows
+def test_can_set_locale_invalid_get(monkeypatch):
+    # In some cases, an invalid locale can be set,
+    # but a subsequent getlocale() raises a ValueError
+    # See GH 22129
+
+    def mocksetlocale(category, locale):
+        return locale
+
+    def mockgetlocale():
+        raise ValueError()
+
+    with monkeypatch.context() as m:
+        m.setattr(locale, 'setlocale', mocksetlocale)
+        m.setattr(locale, 'getlocale', mockgetlocale)
```

Comment:
Do we need to patch this? From the comment above, shouldn't it naturally raise a ValueError after setting `non-existent_locale` as the locale?

Reply:
If locale.Error is raised while setting `non-existent_locale`, then the locale is not changed, and 'getlocale' will return the current locale.

The second test needs to simulate successfully setting `non-existent_locale` **and** raising a ValueError so we can test the changes in `pandas/util/testing.py`.

Reply:
Hmm so does `non-existent_locale` fail when getting or setting? Reason I ask is because I think there's a better way to isolate the expected behavior here, i.e. we should only need to mock the getter or setter but not both within a given test

Reply:
Right - so isn't the patch to `setlocale` unnecessary? So long as `getlocale` raises (as your patch causes to happen) the set behavior is more or less irrelevant (?)

Reply:
If we don't patch `setlocale`, then it will raise locale.Error and we don't get to test what happens when `getlocale` raises ValueError.

Reply:
Only because we are setting an invalid locale though right? If we set the default locale yet mocked the get aspect to it wouldn’t that give the desired effect?

Reply:
Yes, for the second test we could choose a valid locale, and only patch `getlocale` to raise ValueError, but at this point, how do we know the name of a valid locale?

Reply:
Can you call getlocale prior to the patch and subsequently send that as an argument to can_set_locale?

Reply:
On my system (Ubuntu 18.04), Python2 gives me:
```
>>> import locale; locale.getlocale()
(None, None)
```
Whereas Python3 gives me:
```
>>> import locale; locale.getlocale()
('en_ZA', 'UTF-8')
```

Reply:
This works (in Python2 and Python3):
```
>>> locale.setlocale(locale.LC_ALL, '')
'en_ZA.UTF-8'
```
So shall we rather go with `assert tm.can_set_locale('') is False` and drop the monkeypatch on `setlocale`?

Reply:
Are you saying `tm.can_set_locale(locale.getlocale())` isn't working across both versions? I believe a tuple is a valid argument. Would be cleaner to just send that value through instead of the empty string

Reply:
The documentation gives this an example.
https://docs.python.org/2/library/locale.html 
```
import locale
locale.setlocale(locale.LC_ALL, '')

```

Reply:
Ah OK that's fair - thanks for pointing there!